### PR TITLE
LUCENE-10382: Ensure kNN filtering works with other codecs

### PR DESCRIPTION
**Pull request by jtibshirani**
The original PR that added kNN filtering support overlooked non-default codecs.
This follow-up ensures that other codecs work with the new filtering logic:
* Make sure to check the visited nodes limit in `SimpleTextKnnVectorsReader`
and `Lucene90HnswVectorsReader`
* Add a test `BaseKnnVectorsFormatTestCase` to cover this case
* Fix failures in `TestKnnVectorQuery#testRandomWithFilter`, whose assumptions
don't hold when SimpleText is used

This PR also clarifies the limit checking logic for
`Lucene91HnswVectorsReader`. Now we always check the limit before visiting a
new node, whereas before we only checked it in an outer loop.

### CODE REVIEW

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java [hunk 1]
```diff
@@ -821,6 +822,64 @@ public void testRandom() throws Exception {
...
+          }
+
+          // check the limit is hit when it's very small
+          int k = 5 + random().nextInt(45);
+          int visitedLimit = k + random().nextInt(5);
+          TopDocs results =
+              ctx.reader()
+                  .searchNearestVectors(
+                      fieldName, randomVector(dimension), k, liveDocs, visitedLimit);
+          assertEquals(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO, results.totalHits.relation);
```
Comment (jpountz) [inline lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:866]: Would this be actually guaranteed for every KNN vectors format? It feels hard to identify nearest neighbors while visitedLimit is so close to k, but maybe some edge cases like completely disconnected graphs could trigger rare test failures?

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java [hunk 2]
```diff
@@ -155,18 +159,13 @@ private NeighborQueue searchLevel(
     if (results.size() >= topK) {
       bound.set(results.topScore());
     }
-    while (candidates.size() > 0) {
+    while (candidates.size() > 0 && results.incomplete() == false) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:162]: Maybe extract this to an outer `if` statement so that we only check it before entering the while loop. Once we are in the while loop, `results.incomplete()` is guaranteed to be `false` since we always break the loop after calling `markIncomplete()`?

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java [hunk 3]
```diff
@@ -186,6 +188,7 @@ private NeighborQueue searchLevel(
             }
           }
         }
+        numVisited++;
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:191]: nit: I liked having `numVisited` next to `similarityFunction.compare` better since my mental model is that the point of `numVisited` is to compute the number of similarity computations that are performed to find the top-k nearest neighbors to the query vector.

**Review by jpountz — APPROVED** [review]

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java (hunk 3 above)
Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:191]: I started to think of visiting as both performing a similarity calculation and potentially updating the queue. I'm happy to move it back there though, your perspective makes sense too.

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java [hunk 4]
```diff
@@ -821,6 +822,64 @@ public void testRandom() throws Exception {
...
 
+  public void testSearchWithVisitedLimit() throws Exception {
+    IndexWriterConfig iwc = newIndexWriterConfig();
+    String fieldName = "field";
+    try (Directory dir = newDirectory();
+        IndexWriter iw = new IndexWriter(dir, iwc)) {
+      int numDoc = atLeast(300);
+      int dimension = atLeast(10);
+      for (int i = 0; i < numDoc; i++) {
+        int id = random().nextInt(numDoc);
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:833]: Oops, I did not mean to do this (copied this from another test).

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java (hunk 1 above)
Comment (jtibshirani) [reply lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:866]: Right, this isn't a behavioral guarantee. I was seeing this test just as providing important signal (catching if you forgot to implement it, making sure it works in non-edge case scenarios).

I ran it a ton of times locally and didn't see failures. I'll keep an eye on builds -- if we encounter a failure because of the random data generation, an option is to switch to using a fixed set of vectors.

Comment (jpountz) [reply lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:866]: This sounds ok to me, I was mostly curious to better understand how you were thinking of it. We can revisit when/if we see failures with a new codec or other changes. It's not the only place where we are doing things like that. :)

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java (hunk 2 above)
Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:162]: The 'break' only breaks out of the inner while loop, which checks the neighbors list. This check is required so that when that break happens, we exit the outer loop too. I preferred to do this rather than using a labelled while statement, which can catch people by surprise. Did I understand your comment correctly?

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:162]: Ah, thanks for clarifying, I had missed that the break was now under the inner while loop. Let's keep things as they are in your PR.

**Review by jpountz — APPROVED** [review]
